experimental/parser/lexer.go

```diff
@@ -0,0 +1,675 @@
+				return !unicode.IsPrint(r)
+			})
+			if id == "" {
+				// This "identifier" appears to consist entirely of unprintable
+				// characters (e.g. combining marks).
+				tok := l.Push(len(rawId), token.Unrecognized)
+				l.Error(ErrUnrecognized{Token: tok})
+				continue
+			}
+
+			l.cursor -= len(rawId) - len(id)
+			tok := l.Push(len(id), token.Ident)
+
+			// Legalize non-ASCII runes.
+			if !isASCIIIdent(tok.Text()) {
+				l.Error(ErrNonASCIIIdent{Token: tok})
+			}
+
+		default:
+			// Back up behind the rune we just popped.
+			l.cursor -= utf8.RuneLen(r)
+
+			// Consume as many grapheme clusters as possible, and diagnose it.
+			unknown := l.TakeGraphemesWhile(func(g string) bool {
+				r, _ := utf8.DecodeRuneInString(g)
+				return !strings.ContainsRune(";,/:=-.([{<>}])_\"'", r) &&
+					!xidContinue(r) &&
+					!unicode.In(r, unicode.Pattern_White_Space)
+			})
+			tok := l.Push(len(unknown), token.Unrecognized)
+			l.Error(ErrUnrecognized{Token: tok})
+		}
+	}
+
+	// Legalize against unclosed delimiters.
+	for _, open := range l.openStack {
+		l.Error(ErrUnterminated{Span: open.Span()})
+	}
+	// In backwards order, generate empty tokens to fuse with
+	// the unclosed delimiters.
+	for i := len(l.openStack) - 1; i >= 0; i-- {
+		empty := l.Push(0, token.Unrecognized)
+		token.Fuse(l.openStack[i], empty)
+	}
+
+	// Perform implicit string concatenation.
+	catStrings := func(start, end token.Token) {
+		var buf strings.Builder
+		for i := start.ID(); i <= end.ID(); i++ {
+			tok := i.In(l.Context)
+			if s, ok := tok.AsString(); ok {
+				buf.WriteString(s)
+				token.ClearValue(tok)
+			}
+		}
+		token.SetValue(start, buf.String())
+		token.Fuse(start, end)
+	}
+	var start, end token.Token
+	l.All()(func(tok token.Token) bool {
+		if tok.IsSynthetic() {
+			return false
+		}
+
+		switch tok.Kind() {
+		case token.Space, token.Comment:
+			break
+		case token.String:
+			if start.Nil() {
+				start = tok
+			} else {
+				end = tok
+			}
+		default:
+			if !start.Nil() && !end.Nil() {
+				catStrings(start, end)
+			}
+			start = token.Nil
+			end = token.Nil
+		}
+		return true
+	})
+	if !start.Nil() && !end.Nil() {
+		catStrings(start, end)
+	}
+}
+
+func (l *lexer) Push(length int, kind token.Kind) token.Token {
+	l.count++
+	return l.Stream.Push(length, kind)
+}
+
+// LexString lexes a number starting at the current cursor.
+func (l *lexer) LexNumber() token.Token {
+	start := l.cursor
+	// Accept all digits, legalize later.
+	// Consume the largest prefix that satisfies the rules at
+	// https://protobuf.com/docs/language-spec#numeric-literals
+more:
+	r := l.Peek()
+	if r == 'e' || r == 'E' {
+		_ = l.Pop()
+		r = l.Peek()
+		if r == '+' || r == '-' {
+			_ = l.Pop()
+		}
+
+		goto more
+	}
+	if r == '.' || unicode.IsDigit(r) || unicode.IsLetter(r) ||
+		// We consume _ because 0_n is not valid in any context, so we
+		// can offer _ digit separators as an extension.
+		r == '_' {
+		_ = l.Pop()
+		goto more
+	}
+
```

Comment:
It seems like after this point, we never use a `goto`, so a `for { }` loop above could just be the above statements (with a `break` at the end of the body; or invert the conditions and break instead of continue).